Xming: Lock X display during winClipboardFlushXEvents event processing

XXX: As far as I can see, these are only locking over a single Xlib call,
so it's a bit unclear what race condition that can be preventing...


Copyright (C) Colin Harrison 2005-2008
http://www.straightrunning.com/XmingNotes/
http://sourceforge.net/projects/xming/

---
 xserver/hw/xwin/winclipboardxevents.c |    8 ++++++++
 1 file changed, 8 insertions(+)

Index: xorg-server-1.5.3/xserver/hw/xwin/winclipboardxevents.c
===================================================================
--- xorg-server-1.5.3.orig/xserver/hw/xwin/winclipboardxevents.c	2009-02-10 21:56:31.218750000 +0000
+++ xorg-server-1.5.3/xserver/hw/xwin/winclipboardxevents.c	2009-02-10 21:56:38.828125000 +0000
@@ -188,6 +188,7 @@
 	    }
 
 	  /* Check that clipboard format is available */
+	  XLockDisplay (pDisplay);
 	  if (fUseUnicode
 	      && !IsClipboardFormatAvailable (CF_UNICODETEXT))
 	    {
@@ -200,6 +201,7 @@
 	      lasthwnd = hwnd;
 
 	      /* Abort */
+	      XUnlockDisplay (pDisplay);
 	      fAbort = TRUE;
 	      goto winClipboardFlushXEvents_SelectionRequest_Done;
 	    }
@@ -210,6 +212,7 @@
 		      "available from Win32 clipboard.  Aborting.\n");
 
 	      /* Abort */
+	      XUnlockDisplay (pDisplay);
 	      fAbort = TRUE;
 	      goto winClipboardFlushXEvents_SelectionRequest_Done;
 	    }
@@ -228,6 +231,7 @@
 		      GetLastError ());
 
 	      /* Abort */
+	      XUnlockDisplay (pDisplay);
 	      fAbort = TRUE;
 	      goto winClipboardFlushXEvents_SelectionRequest_Done;
 	    }
@@ -269,6 +273,7 @@
 		      GetLastError ());
 
 	      /* Abort */
+	      XUnlockDisplay (pDisplay);
 	      fAbort = TRUE;
 	      goto winClipboardFlushXEvents_SelectionRequest_Done;
 	    }
@@ -304,6 +309,7 @@
 
 	  /* Convert DOS string to UNIX string */
 	  winClipboardDOStoUNIX (pszConvertData, strlen (pszConvertData));
+	  XUnlockDisplay (pDisplay);
 
 	  /* Setup our text list */
 	  pszTextList[0] = pszConvertData;
@@ -675,6 +681,7 @@
 	  xtpText.value = NULL;
 
 	  /* Convert the X clipboard string to DOS format */
+	  XLockDisplay (pDisplay);
 	  winClipboardUNIXtoDOS (&pszReturnData, strlen (pszReturnData));
 
 	  if (fUseUnicode)
@@ -782,6 +789,7 @@
 	   */
 
 	winClipboardFlushXEvents_SelectionNotify_Done:
+	  XUnlockDisplay (pDisplay);
 	  /* Free allocated resources */
 	  if (ppszTextList)
 	    XFreeStringList (ppszTextList);
